### PR TITLE
[FEATURE] Allow files to be references

### PR DESCRIPTION
**Pull request by e0ipso**
Allow files to be treated as entity references. Instead of getting
the file array, get the entity id or the full entity. To use this
behaviour use the 'class' property in the publicFields() definition.

### CODE REVIEW

**Comment by e0ipso** [comment]
This PR depends on #592. Please, merge that one first.

**Comment by amitaibu** [comment]
@mateu-aguilo-bosch can you please update the different PRs with latest code - for easier review.

**Comment by e0ipso** [comment]
@amitaibu done.

File: src/Plugin/resource/Field/ResourceFieldFileEntityReference.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+   *
+   * @return \EntityMetadataWrapper
+   *   Either a \EntityStructureWrapper or a \EntityListWrapper.
+   *
+   * @throws ServerConfigurationException
+   */
+  protected function propertyWrapper(DataInterpreterInterface $interpreter) {
+    $property_wrapper = parent::propertyWrapper($interpreter);
+    // If the file_entity module is not installed, throw an exception.
+    if (!module_exists('file_entity')) {
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldFileEntityReference.php:30]: why? You can use the wrapper on a file, without file_entity. AFAIK file_entity just makes that entity fieldable.

File: src/Plugin/resource/Field/ResourceFieldEntity.php [hunk 2]
```diff
@@ -440,6 +445,27 @@ protected function formatterValue(\EntityMetadataWrapper $property_wrapper, \Ent
   }
 
   /**
+   * Gets the list of fields to show in the embedded resource.
+   *
+   * @return string[]
+   *   The list of fields.
+   */
+  protected function emdeddedSparseFieldsets() {
+    // TODO: Inject the request object here.
```
Comment (amitaibu) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:454]: isn't it just doing `emdeddedSparseFieldsets(RequestInterface ..)`?

**Comment by amitaibu** [comment]
minor comments, other comments done on another mega PR

File: src/Plugin/resource/Field/ResourceFieldFileEntityReference.php (hunk 1 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldFileEntityReference.php:30]: Good point. I don't remember exactly why I put this (maybe entity access related?). Probably safe to remove.

File: src/Plugin/resource/Field/ResourceFieldEntity.php (hunk 2 above)
Comment (e0ipso) [inline src/Plugin/resource/Field/ResourceFieldEntity.php:454]: The problem is that we don't have the request object in resource fields. Good news is that I (think that I) added support for that in a subsequent PR.

**Comment by e0ipso** [comment]
Merged.

**Pull request closed without merging**